### ISSUE TITLE
Feature request: flowtype the library

### ISSUE THREAD

**Issue by gre** [paularmstrong/normalizr#309]
I've been using normalizr for a month now and it helps us a lot building an API/data library. The one thing that I find missing is it is difficult to have the objects typed properly regarding the schema provided by JS code.

More simply, I imagine we could at least have this possible:

```js
//@flow

class UserSchema extends schema.Entity<{ id: string, name: string, age: number }> {}

const userSchema = new UserSchema();
const mySchema = { users: [ user ] }
const entities = { users: { '1': { id: '1', name: "gre", age: 28 } };
const denormalizedData = denormalize({ users: [ 1 ] }, mySchema, entities);

// denormalizedData type inferred to `{ users: Array<{ id: string, name: string, age: number }> }`
```

NB: i'm not sure an instance can hold a parametric type, therefore I have made class for UserSchema in this example.
NB: more advanced usecase is obviously to make this work when you compose schemas, I imagine you could also express this in flowtype so things get even more inferred so user have less code to write. The only thing we can't really infer is the field that are not part of the normalization ;)

Cheers

**Comment by dvlsg** [comment]
Not sure how reliable it is since I typically just use typescript, myself, but you could try sending the existing `index.d.ts` through [flowgen](https://github.com/joarwilk/flowgen).

**Comment by gre** [comment]
interesting, I didn't know there were a typescript def :)

**Comment by TroySchmidt** [comment]
Yeah I wouldn't recommend Flow at all to any project at this point.  TypeScript I think has won the battle and is far more feature rich than Flow.

**Comment by callmetwan** [comment]
>TypeScript I think has won the battle

@TroySchmidt Do you have anything to back that statement up? Flow is really popular in the React community. I'd say declaring a "winner" is premature, TypeScript has simply been around longer. Flow has been improved upon pretty rapidly over it's short lifetime and with FB's backing I only see it getting stronger.

**Comment by TroySchmidt** [comment]
@callmetwan I started in React about two years ago.  At that point, I took the pulse by reading sites, examples, etc and they seemed pretty close in supporting needs and implementations.  I went with Flow and it was sufficient.  As I got in to more complex typing needs using Flow to support those got harder and harder to come up with.  TypeScript matured in the mean time.  So I switched over to TypeScript.  I can say the experience is so much better.
TypeScript support is baked directly into VS Code.  CodeLens implementation and reference counts for interfaces as well as the Intellisense support completely obliterate Flow.  There are many more features that simply because of what Flow is it wouldn't be able to keep up with.  Not to mention my computer fans aren't whirring away trying to continuously keep running it.
Also more and more packages have their TypeScript types directly in their libraries and don't need external definitions or @types libraries.
In practice of building sites, TypeScript is far superior a product to Flow.  Flow might still have a place in packages and libraries.  But in websites built using those packages Flow doesn't compare.

**Comment by paularmstrong** [comment]
Regardless of preferences, I'd like to have flowtype added to Normalizr. Let's not diverge this issue into a discussion of the merits of one versus the other—They're both great.

**Comment by callmetwan** [comment]
I can't argue with that!

**Comment by wuweiweiwu** [comment]
@paularmstrong Can I start working on this? :)

**Comment by re-gor** [comment]
I've made a [pull-request](https://github.com/flow-typed/flow-typed/pull/2717) in flow-typed with declarations for normalizr. Until it is accepted (if accepted), they can be connected to the project via flow-typed/custom. Hope someone find it usefull

**Comment by pkyeck** [comment]
Now that there are flow-declarations, it would be nice to have a supersimple "how-to type" example - maybe for the example in the readme...!?

```javascript
import { normalize, schema } from 'normalizr';

// Define a users schema
const user = new schema.Entity('users');

// Define your comments schema
const comment = new schema.Entity('comments', {
  commenter: user
});

// Define your article
const article = new schema.Entity('articles', {
  author: user,
  comments: [comment]
});

const normalizedData = normalize(originalData, article);
```

flow keeps complaining, that 

    Missing type annotation for `Collections`. `Collections` is a type parameter declared in  function type [1] and was implicitly instantiated at  call of `normalize` [2].

but I don't know where/what to type ...

**Comment by pkyeck** [comment]
@re-gor maybe you can help me with the above question?
